src/main.cc

```diff
@@ -78,10 +74,63 @@ int main(int argc, const char *argv[])
         blueprint.report.error = snowcrash::Error(e.what(), snowcrash::ApplicationError);
     }
 
-    delete out;
-    delete serializer;
-
     PrintReport(blueprint.report, inputStream.str(), config.lineNumbers);
 
     return blueprint.report.error.code;
 }
+
+int ProcessRefract(const Config& config, std::unique_ptr<std::istream>& in, std::unique_ptr<std::ostream>& out) {
+
+    std::stringstream inputStream;
+    inputStream << in->rdbuf();
+
+    drafter_options options;
+    options.sourcemap = config.sourceMap;
+    options.format = config.format == drafter::YAMLFormat
+       ? DRAFTER_SERIALIZE_YAML
+       : DRAFTER_SERIALIZE_JSON;
+
+    refract::IElement* result = nullptr;
+
+    int ret = drafter_parse_blueprint(inputStream.str().c_str(), &result);
+
+    if (!result) {
+        return -1;
+    }
+
+    if (!config.validate) { // If not validate, we serialize
+        char* output = drafter_serialize(result, options);
+
+        if (output) {
+            *out << output
+                 << "\n"
+                 << std::flush;
+
+            free(output);
+        }
+    }
+
+    PrintReport(result, inputStream.str(), config.lineNumbers, ret);
+
+    delete result;
```

Comment:
If one day `drafter_free_result` is more complicated then simple `delete` we would need to fix all of those. Another reason is that this code is an example of usage of the API and it should be the best practice so using the API properly will help others.